### PR TITLE
Improve error handling in SageMaker deployment operator

### PR DESCRIPTION
**Pull request by yubozhao**
Better handling of Sagemaker errors

### CODE REVIEW

File: bentoml/deployment/sagemaker/__init__.py [hunk 1]
```diff
@@ -223,6 +225,44 @@ def __exit__(self, exc_type, exc_val, exc_tb):
 }
 
 
+def parse_aws_client_exception_or_raise(e):
```
Comment (parano) [inline bentoml/deployment/sagemaker/__init__.py]: nit: add prefix? '_parse_aws_client_exception_or_raise'

Comment (yubozhao) [reply bentoml/deployment/sagemaker/__init__.py]: yeah, good call

File: bentoml/deployment/sagemaker/__init__.py [hunk 2]
```diff
@@ -223,6 +225,44 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+            return Status.NOT_FOUND(error_response.get('Message', 'Unknown'))
+        elif error_code == 'InvalidSignatureException':
+            return Status.UNAUTHENTICATED(error_response.get('Message', 'Unknown'))
+        else:
+            raise e
+    elif type(e) == ParamValidationError:
+        return Status.INVALID_ARGUMENT(str(e))
+
+
+def _cleanup(client, name, version, cleanup_endpoint_config=False):
```
Comment (parano) [inline bentoml/deployment/sagemaker/__init__.py]: separate this to `_cleanup_sagemaker_endpoint_config` and  `_cleanup_sagemaker_model`

File: bentoml/deployment/sagemaker/__init__.py [hunk 3]
```diff
@@ -223,6 +225,44 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+        endpoint_config_name = create_sagemaker_endpoint_config_name(name, version)
+        try:
+            delete_endpoint_config_response = client.delete_endpoint_config(
+                EndpointConfigName=endpoint_config_name
+            )
+            logger.debug(
+                "AWS delete endpoint config response: %s",
+                delete_endpoint_config_response,
+            )
+        except (ClientError, ParamValidationError) as e:
```
Comment (parano) [inline bentoml/deployment/sagemaker/__init__.py]: as discussed - remove ParamValidationError

File: bentoml/deployment/sagemaker/__init__.py [hunk 4]
```diff
@@ -284,32 +345,99 @@ def apply(self, deployment_pb, repo, prev_deployment=None):
...
         endpoint_name = generate_aws_compatible_string(
             deployment_pb.namespace + '-' + deployment_spec.bento_name
         )
-        if prev_deployment:
-            logger.info("Updating sagemaker endpoint %s", endpoint_name)
-            update_endpoint_response = sagemaker_client.update_endpoint(
-                EndpointName=endpoint_name, EndpointConfigName=endpoint_config_name
+        try:
+            if prev_deployment:
+                logger.info("Updating sagemaker endpoint %s", endpoint_name)
```
Comment (parano) [inline bentoml/deployment/sagemaker/__init__.py]: use logger.debug ?

File: bentoml/deployment/sagemaker/__init__.py [hunk 5]
```diff
@@ -284,32 +345,99 @@ def apply(self, deployment_pb, repo, prev_deployment=None):
...
+                update_endpoint_response = sagemaker_client.update_endpoint(
+                    EndpointName=endpoint_name, EndpointConfigName=endpoint_config_name
+                )
+                logger.debug(
+                    "AWS update endpoint response: %s", update_endpoint_response
+                )
+            else:
+                logger.info("Creating sagemaker endpoint %s", endpoint_name)
```
Comment (parano) [inline bentoml/deployment/sagemaker/__init__.py]: use logger.debug ?

File: bentoml/deployment/sagemaker/__init__.py [hunk 6]
```diff
@@ -223,6 +226,55 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+
+    Args:
+        e: ClientError from botocore.exceptions
+    Returns:
+        StatusProto
+    """
+    error_response = e.response.get('Error', {})
+    error_code = error_response.get('Code')
+    if error_code == 'ValidationException':
+        logger.info(str(error_response))
```
Comment (parano) [inline bentoml/deployment/sagemaker/__init__.py]: add more information to the log here? e.g. "AWS Client Error: {code} - {message}"

File: bentoml/deployment/sagemaker/__init__.py [hunk 7]
```diff
@@ -223,6 +226,55 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+        return Status.NOT_FOUND(error_response.get('Message', 'Unknown'))
+    elif error_code == 'InvalidSignatureException':
+        logger.info(str(error_response))
+        return Status.UNAUTHENTICATED(error_response.get('Message', 'Unknown'))
```
Comment (parano) [inline bentoml/deployment/sagemaker/__init__.py]: `else raise e` is still necessary here, to make sure when it gets an AWS client error that we don't handle properly

File: bentoml/deployment/sagemaker/__init__.py [hunk 8]
```diff
@@ -223,6 +226,55 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+    Returns:
+        StatusProto
+    """
+    error_response = e.response.get('Error', {})
+    error_code = error_response.get('Code')
+    if error_code == 'ValidationException':
+        logger.info(str(error_response))
+        return Status.NOT_FOUND(error_response.get('Message', 'Unknown'))
+    elif error_code == 'InvalidSignatureException':
+        logger.info(str(error_response))
```
Comment (parano) [inline bentoml/deployment/sagemaker/__init__.py]: similarly here, log more details

**Pull request merged**